From 6cdfc371cd839959ee4a1e54d610c5aa0755d868 Mon Sep 17 00:00:00 2001
From: Rui Loura <rui@joyent.com>
Date: Tue, 23 Jul 2019 16:44:27 +0000
Subject: [PATCH] MANTA-4428 rust-sharkspotter should put each object on a
 newline in default output file MANTA-4492 pin rust-sharkspotter to
 rust-libmanta 0.1.0

---
 Cargo.toml  |  2 +-
 README.md   |  9 +++++++++
 src/lib.rs  | 44 ++++++++++++++++++++------------------------
 src/main.rs |  6 ++----
 4 files changed, 32 insertions(+), 29 deletions(-)

diff --git a/Cargo.toml b/Cargo.toml
index ccdf20e..d397b6e 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -9,7 +9,7 @@ assert_cli = "0.6.0"
 serde = { version = "1.0.89", features = ["derive"] }
 serde_json = "1.0.39"
 clap = "2.33.0"
-libmanta = { git = "https://github.com/joyent/rust-libmanta" }
+libmanta = { git = "https://github.com/joyent/rust-libmanta", tag = "v0.1.0" }
 moray = { git = "https://github.com/joyent/rust-moray"}
 slog = { version = "2.5.2" }
 slog-bunyan = { git = "https://github.com/kellymclaughlin/bunyan", branch = "build-on-smartos" }
diff --git a/README.md b/README.md
index 7fd06bd..a03becb 100644
--- a/README.md
+++ b/README.md
@@ -28,11 +28,20 @@ OPTIONS:
     -s, --shark <STORAGE_ID>          Find objects that belong to this shark
 ```
 
+## Example
+```
+$ cargo run -- --domain east.joyent.us --shark 1.stor -M 1 -m 1
+$ json -f shard_1_1.stor.objs -g
+```
+
+__This can be a big file so [json](https://github.com/trentm/json) may struggle with it__
+
 ## Development
 
 Before integration run:
 ```
 cargo fmt
+cargo clippy
 ```
 
 and 
diff --git a/src/lib.rs b/src/lib.rs
index a2e21a2..a5800a7 100644
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -80,26 +80,24 @@ where
     Ok(())
 }
 
+fn chunk_query(id_name: &str, begin: u64, end: u64, count: u64) -> String {
+    format!(
+        "SELECT * FROM manta WHERE {} >= {} AND \
+         {} <= {} AND type = 'object' limit {};",
+        id_name, begin, id_name, end, count
+    )
+}
+
 fn read_chunk<F>(
     mclient: &mut MorayClient,
-    id_name: &str,
-    begin: u64,
-    end: u64,
-    count: u64,
+    query: &str,
     shard_num: u32,
     handler: &mut F,
 ) -> Result<(), Error>
 where
     F: FnMut(MantaObject, u32) -> Result<(), Error>,
 {
-    let query = format!(
-        "SELECT * FROM manta WHERE {} >= {} AND \
-         {} <= {} AND type = 'object' limit {};",
-        id_name, begin, id_name, end, count
-    );
-
-    match mclient.sql(query.as_str(), vec![], r#"{"timeout": 10000}"#, |a| {
-        let id: String = serde_json::from_value(a[0][id_name].clone()).unwrap();
+    match mclient.sql(query, vec![], r#"{"timeout": 10000}"#, |a| {
         query_handler(a, shard_num, handler)
     }) {
         Ok(()) => Ok(()),
@@ -112,7 +110,7 @@ where
 
 fn iter_ids<F>(
     id_name: &str,
-    moray_socket: &String,
+    moray_socket: &str,
     conf: &config::Config,
     log: Logger,
     shard_num: u32,
@@ -121,7 +119,7 @@ fn iter_ids<F>(
 where
     F: FnMut(MantaObject, u32) -> Result<(), Error>,
 {
-    let mut mclient = MorayClient::from_str(moray_socket.as_str(), log, None).unwrap();
+    let mut mclient = MorayClient::from_str(moray_socket, log, None)?;
     let mut start_id = conf.begin;
     let mut end_id = conf.begin + conf.chunk_size - 1;
     let largest_id = match find_largest_id_value(&mut mclient, id_name) {
@@ -139,15 +137,9 @@ where
     }
 
     while remaining > 0 {
-        match read_chunk(
-            &mut mclient,
-            id_name,
-            start_id,
-            end_id,
-            conf.chunk_size,
-            shard_num,
-            &mut handler,
-        ) {
+        let query = chunk_query(id_name, start_id, end_id, conf.chunk_size);
+        match read_chunk(&mut mclient, query.as_str(), shard_num, &mut handler)
+        {
             Ok(()) => (),
             Err(e) => return Err(e),
         };
@@ -173,7 +165,11 @@ where
     Ok(())
 }
 
-pub fn run<F>(conf: &config::Config, log: Logger, mut handler: F) -> Result<(), Error>
+pub fn run<F>(
+    conf: &config::Config,
+    log: Logger,
+    mut handler: F,
+) -> Result<(), Error>
 where
     F: FnMut(MantaObject, u32) -> Result<(), Error>,
 {
diff --git a/src/main.rs b/src/main.rs
index ad1d783..293d90d 100644
--- a/src/main.rs
+++ b/src/main.rs
@@ -1,9 +1,6 @@
 // Copyright 2019 Joyent, Inc.
 
-extern crate sharkspotter;
-
-
-use slog::{o, Logger, Drain};
+use slog::{o, Drain, Logger};
 use std::collections::HashMap;
 use std::env;
 use std::fs::OpenOptions;
@@ -49,6 +46,7 @@ fn main() -> Result<(), Error> {
         let file = file_map.get_mut(&shard).unwrap();
         let buf = serde_json::to_string(&mobj)?;
         file.write_all(buf.as_bytes())?; // TODO: match
+        file.write_all(b"\n")?;
 
         Ok(())
     })
-- 
2.21.0

